predictive_models/predictive_model.py

```diff
@@ -0,0 +1,245 @@
+    def _get_num_features(self):
+        nfeatures = 0
+
+        for cdict in self.metadata['columns']:
+            data_type = cdict['type']
+            attr_name = cdict['name']
+
+            if attr_name != self.labelCol:
+                if data_type == FLOAT or data_type == INTEGER:
+                    nfeatures += 1
+
+                elif data_type == CATEGORICAL or data_type == ORDINAL:
+                    nfeatures += len(cdict['i2s'])
+
+                else:
+                    raise ValueError(f'Unkown data type {data_type} for attribute {attr_name}')
+
+        return nfeatures
+
+    def _get_feature_names(self):
+        featureNames = []
+
+        for i, cdict in enumerate(self.metadata['columns']):
+            data_type = cdict['type']
+            attr_name = cdict['name']
+
+            if attr_name != self.labelCol:
+                if data_type == FLOAT or data_type == INTEGER:
+                    featureNames.append(attr_name)
+
+                elif data_type == CATEGORICAL or data_type == ORDINAL:
+                    col_cats = cdict['i2s']
+                    featureNames.extend([f'{attr_name}_{c}' for c in col_cats])
+
+        return featureNames
+
+    def _impute_missing_values(self, df):
+        dfImpute = df.copy()
+
+        catCols = []
+        numCols = []
+
+        for col in self.metadata['columns']:
+            if col['name'] in list(dfImpute):
+                if col['type'] in [CATEGORICAL, ORDINAL]:
+                    catCols.append(col['name'])
+                elif col['type'] in NUMERICAL:
+                    numCols.append(col['name'])
+
+        self.ImputerCat.fit(df[catCols])
+        dfImpute[catCols] = self.ImputerCat.transform(df[catCols])
+
+        self.ImputerNum.fit(df[numCols])
+        dfImpute[numCols] = self.ImputerNum.transform(df[numCols])
+
+        return dfImpute
+
+    def _one_hot(self, col_data, categories):
+        col_data_onehot = zeros((len(col_data), len(categories)))
+        cidx = [categories.index(c) for c in col_data]
+        col_data_onehot[arange(len(col_data)), cidx] = 1
+
+        return col_data_onehot
+
+
+class ClassificationTask(PredictiveModel):
+    """ A binary or multiclass classification model. """
+
+    def __init__(self, Distinguisher, metadata, labelCol):
+        """
+        :param Distinguisher: sklearn.Classifier: A classification model
+        :param metadata: dict: Metadata description
+        :param labelCol: str: Name of the target variable
+        """
+        super().__init__(metadata, labelCol)
+        self.Distinguisher = Distinguisher
+
+        labels = self._get_labels()
+        self.labels = {l:i for i, l in enumerate(labels)}
+        self.labelsInv = {i:l for l, i in self.labels.items()}
+
+        self.__name__ = f'{self.Distinguisher.__class__.__name__}{self.labelCol}'
+
+    def train(self, data):
+        assert isinstance(data, self.datatype), f"Model expects input as {self.datatype} but got {type(data)}"
```

Comment:
Assertions should be left to testing purposes, there it is better to test and raise an exception.